Issue738 state parameters for bb-tab-sref

Blackbaud-JohnWard opened the pull request:
@Blackbaud-PatrickOFriel

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/blackbaud/skyux?pullRequest=745) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/blackbaud/skyux?pullRequest=745) before we can accept your contribution.<br/>

Blackbaud-PatrickOFriel commented on js/sky/src/tabsref/tabsref.js:
```diff
@@ -47,4 +41,48 @@
+
+    function checkForAndWrapRefWithStateName(ref, $state) {
+        var newRef,
+            preparsedRef = ref.match(/^\s*({[^}]*})\s*$/);
```
For regex noobs (me) /other folks that might have to maintain this later, can you comment in code to explain what this regex is grabbing at a high level?

Blackbaud-PatrickOFriel commented on js/sky/src/tabsref/tabsref.js:
```diff
@@ -47,4 +41,48 @@
+    }
+
+    function parseAndValidateRef(ref) {
+        var parsedRef = ref.replace(/\n/g, " ").match(/^([^(]+?)\s*(\((.*)\))?$/);
```
Same as above

Blackbaud-PaulCrowder commented on js/sky/src/tabsref/tabsref.js:
```diff
@@ -47,4 +41,48 @@
                 }
             };
         }]);
+
```
Also as a matter of best practice we put all function declarations at the top of the containing scope because it reflects the hoisting that occurs with functions at runtime.

Blackbaud-PatrickOFriel commented:
Hey @Blackbaud-JohnWard looking over the complexity of having to parse the object out of a combined state/params attribute, I'm leaning towards just passing the stateParams in an attribute `bb-tab-sref-params` or something of the sort. That way you have a separate state and state params that you can access independently and not worry about trying to grab one from within the other. Does that make sense?

Blackbaud-JohnWard commented:
@Blackbaud-PatrickOFriel I thought about that.  However, I think my approach is probably the best way for users, because it follows the same syntax as ui-router.  You'll notice in the demo I had to add the parameter to the $stateProvider config as well, as this is a requirement of ui-router for non-url parameters.  It could be weird for consumers to understand both the skyux syntax and the ui-router syntax, instead of just having the same syntax for both.

Also, I don't want to give off any impressions that I'm a regex guru, lol.  I actually took those expressions directly from ui-router's source.  I can definitely add some comments to explain what they are looking for in the string.

Blackbaud-JohnWard commented:
Forgot to mention, the user can still pass an object from their controller with this syntax...same as ui-router.  Note the evaluateAngularExpression method as the params object is treated as an angular expression.

Blackbaud-PatrickOFriel commented:
I see what you mean, you're matching the argument that the `ui-sref` takes; that makes sense. And yeah I saw you were evaluating the object after I made the comment, so that's good. Then I would say make two minor changes: 
  - Add a note about the regex in the source (and maybe a link to the ui-router source file where you found it)
  - Have your functions defined before you use them. In the SKY UX source, we usually try to have the functions declared before they're called, so for consistency's sake, let's move those declarations up a bit.

Thanks!

Blackbaud-PatrickOFriel commented:
@blackbaud-sky-savage retry

Blackbaud-PatrickOFriel commented:
#738

Blackbaud-JohnWard commented:
Thanks for the feedback @Blackbaud-PatrickOFriel and @Blackbaud-PaulCrowder 

I've moved the function declarations up, and added some comments for clarity around the regex matching.

Blackbaud-PatrickOFriel commented:
Looks good, one change suggestion for your demo. In order to demonstrate a stateParam actually affecting the url, can you change the url for tabState2 to something like `url: 'tabsref/tab2/:username'`. That way the url will pick up the state parameter on tab change

Blackbaud-JohnWard commented:
@Blackbaud-PatrickOFriel Thanks for the feedback.  I went ahead and added the url parameter to the demo.  As far as the urlRouterProvider trick, that's probably a bit out of scope for bb-tab-sref.  Angular-ui-router docs can provide more extensive info for users.

Blackbaud-PatrickOFriel commented:
Approved

blackbaud-sky-build-user commented:
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/745/storybook/)

Component Storybooks:

- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/ag-grid-storybook/)
- [indicators-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/indicators-storybook/)
- [layout-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/layout-storybook/)
- [modals-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/modals-storybook/)
- [pages-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/pages-storybook/)
- [popovers-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/popovers-storybook/)
- [split-view-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/split-view-storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/tabs-storybook/)
- [tiles-storybook](https://blackbaud.github.io/skyux-pr-preview/745/storybooks/tiles-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/745/code-examples/)
- [dep-graph](https://blackbaud.github.io/skyux-pr-preview/745/dep-graph/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/745/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/745/playground/)
